cin_validator/rules/cin2022_23/rule_8940.py

```diff
@@ -0,0 +1,262 @@
+    message="Child Protection Plan data contains overlapping dates",
+    # The column names tend to be the words within the < > signs in the github issue description.
+    affected_fields=[
+        CPPstartDate,
+        CPPendDate,
+    ],
+)
+def validate(
+    data_container: Mapping[CINTable, pd.DataFrame], rule_context: RuleContext
+):
+    # PREPARING DATA
+
+    # Replace ChildProtectionPlans with the name of the table you need.
+    df_cpp = data_container[ChildProtectionPlans].copy()
+    df_cpp2 = data_container[ChildProtectionPlans].copy()
+
+    # Before you begin, rename the index so that the initial row positions can be kept intact.
+    df_cpp.index.name = "ROW_ID"
+    df_cpp2.index.name = "ROW_ID"
+
+    # Resetting the index causes the ROW_IDs to become columns of their respective DataFrames
+    # so that they can come along when the merge is done.
+    df_cpp.reset_index(inplace=True)
+    df_cpp2.reset_index(inplace=True)
+
+    # ReferenceDate exists in the header table so we get header table too.
+    df_ref = data_container[Header]
+    ref_date_series = df_ref[ReferenceDate]
+
+    # the make_census_period function generates the start and end date so that you don't have to do it each time.
+    collection_start, reference_date = make_census_period(ref_date_series)
+
+    # lOGIC
+    # Implement rule logic as described by the Github issue.
+    # Put the description as a comment above the implementation as shown.
+
+    # Where more than one <ChildProtectionPlans> group is included,
+    # the <CPPstartDate> (N00105) of each group cannot fall within either:
+    #   a) <CPPstartDate> (N00105) to <CPPendDate> (N00115), or
+    #   b) <CPPstartDate> (N00105) and <ReferenceDate> if <CPPendDate> (N00115) is not present
+    # of any other group
+    #
+    # Issues dfs should return rows where CPPstartDate is between another CPPstartDate and CPPendDate (or ReferenceDate)
+
+    #  Create dataframes which only have rows with CP plans, and which should have one plan per row.
+    df_cpp = df_cpp[df_cpp[CPPstartDate].notna()]
+    df_cpp2 = df_cpp2[df_cpp2[CPPstartDate].notna()]
+    print("\n df_cpp\n", df_cpp)
+    #  Merge tables to test for overlaps
+    df_merged = df_cpp.merge(
+        df_cpp2,
+        on=["LAchildID"],
+        how="left",
+        suffixes=("_cpp", "_cpp2"),
+    )
+
+    # Exclude rows where the CPPID is the same on both sides
+    df_merged = df_merged[(df_merged["CPPID_cpp"] != df_merged["CPPID_cpp2"])]
+    print("\n df_merged exclude\n", df_merged)
+    # Determine whether CPP overlaps another CPP
+    cpp_started_after_start = (
+        df_merged["CPPstartDate_cpp"] >= df_merged["CPPstartDate_cpp2"]
+    )
+    cpp_started_before_end = (
+        df_merged["CPPstartDate_cpp"] < df_merged["CPPendDate_cpp2"]
+    ) & df_merged["CPPendDate_cpp2"].notna()
+    cpp_started_before_refdate = (
+        df_merged["CPPstartDate_cpp"] < reference_date
+    ) & df_merged["CPPendDate_cpp2"].isna()
+
```

Comment:
Good initiative. I can see your thought process.

One other way of doing this could be to replace all the NaN values in CPPendDate of cpp2 before merging.  You could do that by
`df_cpp2[CppEndDate] = df_cpp2[CppEndDate].fillna(reference_date)`

Reply:
You don't have to change how you've done it. 
Just mentioning.

You can read about fillna [here in the docs](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html) and [here on stackoverflow](https://stackoverflow.com/questions/38134012/pandas-dataframe-fillna-only-some-columns-in-place)